internal/worker_versioning_rules.go

```diff
@@ -0,0 +1,338 @@
+		Percentage float32
+	}
+
+	// VersioningAssignmentRule is a BuildID assigment rule for a task queue.
+	// Assignment rules only affect new workflows.
+	VersioningAssignmentRule struct {
+		// The BuildID of new workflows affected by this rule.
+		TargetBuildId string
+		// A strategy for gradual workflow deployment.
+		Ramp VersioningRamp
+		// An optional registration time set by the server.
+		Timestamp *time.Time
+	}
+
+	// VersioningAssignmentRule is a BuildID redirect rule for a task queue.
+	// It changes the behavior of currently running workflows and new ones.
+	VersioningRedirectRule struct {
+		SourceBuildId string
+		TargetBuildId string
+		// An optional registration time set by the server.
+		Timestamp *time.Time
+	}
+
+	//VersioningConflictToken is a conflict token to serialize updates.
+	// An update with an old token fails with `serviceerror.FailedPrecondition`.
+	// The current token can be obtained with [GetWorkerVersioningRules], or returned by a successful [UpdateWorkerVersioningRules].
+	VersioningConflictToken struct {
+		token []byte
+	}
+
+	// UpdateWorkerVersioningRulesOptions is the input to Client.UpdateWorkerVersioningRules.
+	UpdateWorkerVersioningRulesOptions struct {
+		// The task queue to update the versioning rules of.
+		TaskQueue string
+		// A conflict token to serialize updates.
+		ConflictToken VersioningConflictToken
+		Operation     UpdateVersioningOp
+	}
+
+	// UpdateVersioningOp is an interface for the different operations that can be
+	// performed when updating the worker versioning rules for a task queue.
+	//
+	// Possible operations are:
+	//   - VersioningOpInsertAssignmentRule
+	//   - VersioningOpReplaceAssignmentRule
+	//   - VersioningOpDeleteAssignmentRule
+	//   - VersioningOpInsertRedirectRule
+	//   - VersioningOpReplaceRedirectRule
+	//   - VersioningOpDeleteRedirectRule
+	//   - VersioningOpCommitBuildId
+	UpdateVersioningOp interface {
+		isValidUpdateVersioningOp() bool
+	}
+	VersioningOpInsertAssignmentRule struct {
+		RuleIndex int32
+		Rule      VersioningAssignmentRule
+	}
+	VersioningOpReplaceAssignmentRule struct {
+		RuleIndex int32
+		Rule      VersioningAssignmentRule
+		Force     bool
+	}
+	VersioningOpDeleteAssignmentRule struct {
+		RuleIndex int32
+		Force     bool
+	}
+	VersioningOpInsertRedirectRule struct {
+		Rule VersioningRedirectRule
+	}
+	VersioningOpReplaceRedirectRule struct {
+		Rule VersioningRedirectRule
+	}
+	VersioningOpDeleteRedirectRule struct {
+		SourceBuildId string
+	}
+	VersioningOpCommitBuildId struct {
+		TargetBuildId string
+		Force         bool
+	}
+)
+
+func (uw *UpdateWorkerVersioningRulesOptions) validateAndConvertToProto(namespace string) (*workflowservice.UpdateWorkerVersioningRulesRequest, error) {
+	if namespace == "" {
+		return nil, errors.New("missing namespace argument")
+	}
+	if uw.TaskQueue == "" {
+		return nil, errors.New("missing TaskQueue field")
+	}
+	if !uw.Operation.isValidUpdateVersioningOp() {
```

Comment:
```suggestion
	if err := uw.Operation.validate(); err != nil {
```

This is often a better approach to letting individual implementations validate themselves and return a suitable error. Consider using this over booleans for user error clarity.

Reply:
Yes, the error could add more context on why it didn't validate...